src/Processors/QueryPlan/ReadFromPreparedSource.h

```diff
@@ -21,14 +21,18 @@ class ReadFromPreparedSource : public ISourceStep
     Pipe pipe;
 };
 
-class ReadFromStorageStep : public ReadFromPreparedSource
+class ReadFromStorageStep final : public ReadFromPreparedSource
 {
 public:
-    ReadFromStorageStep(Pipe pipe_, String storage_name, ContextPtr context_, const SelectQueryInfo & query_info_);
+    ReadFromStorageStep(Pipe pipe_, StoragePtr storage_, ContextPtr context_, const SelectQueryInfo & query_info_);
 
     String getName() const override { return "ReadFromStorage"; }
 
+    const StoragePtr & getStorage() const { return storage; }
```

Comment:
@vdimir could you clarify in which case do we need it exactly?

Reply:
This method is required by storage join when, instead of reading data directly from storage, we need to use pre-prepared data for a JOIN operation (e.g., `StorageJoin` or a dictionary).

Currently, this method is unused, as we retrieve storage from the table node in the query tree: https://github.com/ClickHouse/ClickHouse/blob/79225f2e4c2acc4cc75b4eafc579696fcf3087cb/src/Planner/PlannerJoinsLogical.cpp#L389-L398

However, in the future, we plan to fully re-implement this functionality on top of the query plan. I began modifying it, but postponed the changes, leaving this method as part of an earlier attempt.